File: 46-permutations/46-permutations.cpp
class Solution {
    private:
    void findPermutation(vector<int>& nums,vector<int>&ds, int freq[],vector<vector<int>> &ans){
        // base case:
        if(ds.size() == nums.size()){
            ans.push_back(ds);
            return;
        }
        // main logic:
        for(int i= 0;i<nums.size();i++){
            if(!freq[i]){
                freq[i] = 1;
                ds.push_back(nums[i]);
                findPermutation(nums,ds,freq,ans);
                ds.pop_back();
                freq[i] = 0;
            }
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        // Approach_01: Time:n!*n and Space:O(n)+O(n)
        vector<vector<int>> ans;
        vector<int> ds;
        int freq[nums.size()];
        for(int i =0;i<nums.size();i++)
            freq[i] = 0;
        findPermutation(nums,ds,freq,ans);
        return ans;
    }
};

File: 46-permutations/46-permutations.java
class Solution {
    private void findPermutation(int[]nums,List<Integer>ds,boolean []freq,List<List<Integer>> ans){
        // base case:
        if(ds.size() == nums.length){
            ans.add(new ArrayList<>(ds));
            return;
        }
        // main logic:
        for(int i= 0;i<nums.length;i++){
            if(!freq[i]){
                freq[i] = true;
                ds.add(nums[i]);
                findPermutation(nums,ds,freq,ans);
                ds.remove(ds.size()-1);
                freq[i] = false;
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        // Approach_01: Time:n!*n and Space:O(n)+O(n)
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();
        boolean []freq = new boolean[nums.length];
        findPermutation(nums,ds,freq,ans);
        return ans;
    }
}